iris/man/man1/iris-convert.1
.TH iris-convert 1 "February 2011"

.SH NAME
iris-convert

.SH SYNOPSIS 
.B iris-convert
[ --help ] [ --verbose ] [ --setname nodename ] 
[ --clobberName ] [ --normals ] 
[ --static ] 
[ --copyprimitives ] 
[ --nostatesets ] [ --nolighting ] 
[ --optimize [ optimization ] 
infile [ ... ]
outfile

.SH DESCRIPTION

\fBiris-convert\fR reads one or more IRIS readable files, applies the specified conversions, and writes the results to a single output file. The output file type is any OSG writable file.

.SH OPTIONS

.TP
--help
.br
Print the usage message and exit

.TP
--verbose
.br
Sets the dtkMessage spewage level to INFO if the dtkMessage severity is
greater than INFO (such as the default, NOTICE), causing scenegraph
and optimization statistics to be printed to standard error.

.TP
--setname \fInodeName\fR
.br
Sets the name of the node at the top of the returned scenegraph to
\fInodeName\fR if its node name is not set.

.TP
--clobbername 
.br
Always sets the name of the node at the top of the returned
scenegraph.  The --setname parameter is used, or the name of the output
file if the --setname option wasn't given.

.TP
--normals
.br
Calls the osgUtil::SmoothingVisitor to add/replace the normals in the scenegraph

.TP
--static
.br
Changes the datavariance of the scenegraph to STATIC.  This can result in
faster frame rates.

.TP
--copyprimitives
.br
Copy all primitives, as opposed to copying instances of primitives

.TP
--nostatesets
.br
Removes all osg::StateSets from the scenegraph.

.TP
--nolighting
.br
Removes all lighting from the scenegraph.

.TP
--optimize \fIoptimization\fR
.br
Calls iris::Optimizer with the specified options, in the order given. For a
list of optimizations see the osgUtil::Optimizer class reference via the URL
http://www.openscenegraph.org/documentation/OpenSceneGraphReferenceDocs/classes.html.
Optiizations can be passed as either a colon separated list, or a space
separated list if the optimizations are enclosed in quotes so they
parse as a single parameter.

\fBNote that the optimizations are appplied only after the other specified
conversions are applied.\fR

.TP 
infile ...
.br
A list of one or more input files

.TP
outfile
.br
The file to which the scenegraph is written.  Any OSG writable file type is
supported, such as osga or osgb.

.SH NOTE
Setting the DTK_SPEW enviornment variable to info will generate additional
information about the optimizations being performed.

.SH EXAMPLE

This command will convert the file foo.savg to foo.osgb, converting the
data variance of all nodes in the file to STATIC and then applying the
MERGE_GEOMETRY optimization. Additional information about the optimization
will be generated.

env DTK_SPEW=info iris-convert --static --optimize MERGE_GEOMETRY foo.savg foo.osgb

This command will convert the file foo.savg to foo.osgb, applying the default
optimizations.

iris-convert --optimize DEFAULT foo.savg foo.osgb

.SH AUTHOR

.PP
John Kelso, kelso@nist.gov
.br
NIST High Performance Computing and Visualization Group
.br
February 2011
.br
\fBhttp://www.nist.gov/mcsd/hpcvg/\fR
 


iris/src/program/iris-convert.cpp
#include <osgDB/ReadFile>
#include <osgDB/WriteFile>
#include <osgDB/FileUtils>
#include <osgUtil/SmoothingVisitor>
#include <osgUtil/Statistics>

#include <iris.h>

void usage()
{
    fprintf(stderr,"Usage: iris-convert [ --verbose ] [ --help ] [ --setname nodename ] [ --clobberName ] [ --normals ] [ --static ] [ --copyprimitives ] [ --optimize optimization ] [ --nostatesets ] [ --nolighting ] infile [ ... ] outfile\n") ;
}

int main(int argc, char** argv)
{
    if (argc<3)
    {
	usage() ;
	return 1 ;
    }
    // send messages to stderr
    dtkMsg.setFile(stderr) ;

    // use an ArgumentParser object to manage the program arguments.
    iris::ArgumentParser args(&argc,argv);

    bool normals = false ;
    bool setName = false ;
    bool clobberName = false ;
    std::string nodeName ;
    std::string optimizations ;
    std::vector<std::string> load ;

    // for getting rid of statsets, dynamic variance
    iris::ConvertNodeVisitor nv ;

    // parse the options
    int i = 1;
    while (i=args.findOption(i))
    {
	if (iris::IsSubstring("--help",args[i],4))
	{
	    usage() ;
	    return 0 ;
	}
	else if (iris::IsSubstring("--normals",args[i],4))
	{
	    normals = true ;
	    args.remove(i) ;
	}
	else if (iris::IsSubstring("--verbose",args[i],4))
	{
	    if (!dtkMsg.isSeverity(DTKMSG_INFO)) dtkMsg.reset(DTKMSG_INFO) ;
	    args.remove(i) ;
	}
	else if (iris::IsSubstring("--clobbername",args[i],4))
	{
	    clobberName = true ;
	    args.remove(i) ;
	}
	else if (iris::IsSubstring("--setname",args[i],4))
	{
	    if (args.argc()>i+1)
	    {
		setName = true ;
		nodeName = args[i+1] ;
		args.remove(i,2) ;
	    }
	    else
	    {
		dtkMsg.add(DTKMSG_ERROR, "iris-convert: not enough parameters\n") ;
		return 1 ;
	    }
	}
	else if (iris::IsSubstring("--static",args[i],4))
	{
	    nv.setStatic(true) ;
	    args.remove(i) ;
	}
	else if (iris::IsSubstring("--copyprimitives",args[i],4))
	{
	    nv.setCopyPrimitives(true) ;
	    args.remove(i) ;
	}
	else if (iris::IsSubstring("--nostatesets",args[i],5))
	{
	    nv.setNoStateSets(true) ;
	    args.remove(i) ;
	}
	else if (iris::IsSubstring("--nolighting",args[i],5))
	{
	    nv.setNoLighting(true) ;
	    args.remove(i) ;
	}
	else if (iris::IsSubstring("--optimize",args[i],4))
	{
	    if (args.argc()>i+1)
	    {
		optimizations = args[i+1] ;
		args.remove(i,2) ;
	    }
	    else
	    {
		dtkMsg.add(DTKMSG_ERROR, "iris-convert: not enough parameters\n") ;
		return 1 ;
	    }
	}
	else
	{
	    dtkMsg.add(DTKMSG_ERROR, "iris-convert: unknown option \"%s\"\n",args[i]) ;
	    return 1 ;
	}
    }

    if (args.argc()<2)
    {
	dtkMsg.add(DTKMSG_ERROR, "iris-convert: not enough file names\n") ;
	return 1 ;
    }

    osg::ref_ptr<osg::Group> group = new osg::Group ;

    // get the input files
    while(argc>2)
    {
	load.push_back(args[1]) ;	
	args.remove(1) ;
    }

    // load them
    osg::ref_ptr<osg::Node> node = osgDB::readNodeFiles(load);
    if (!node.valid())
    {
	dtkMsg.add(DTKMSG_ERROR, "iris-convert: no model files loaded\n") ;
	return 1 ;
    }
    
    if (normals)
    {
 	dtkMsg.add(DTKMSG_INFO, "iris-convert: adding smooth normals\n") ;
	osgUtil::SmoothingVisitor sv ;
	node->accept(sv) ;
    }

    if (nv.getStatic()) dtkMsg.add(DTKMSG_INFO, "iris-convert: setting DataVariance to static\n") ;
    if (nv.getCopyPrimitives()) dtkMsg.add(DTKMSG_INFO, "iris-convert: copying primitives\n") ;
    if (nv.getNoStateSets()) dtkMsg.add(DTKMSG_INFO, "iris-convert: removing StateSets\n") ;
    if (nv.getNoLighting()) dtkMsg.add(DTKMSG_INFO, "iris-convert: removing lighting\n") ;
    if (nv.getStatic() || nv.getNoStateSets() || nv.getNoLighting() || nv.getCopyPrimitives()) node->accept(nv) ;

    if (nodeName == "") nodeName = args[1] ;

    if (clobberName || (setName && node->getName() == ""))
    {
	dtkMsg.add(DTKMSG_INFO,"iris-convert: setting node name to %s\n",nodeName.c_str()) ;
	node->setName(nodeName) ;
    }

    // optimize last
    if (optimizations.length()>0)
    {
	iris::Optimize(node.get(), optimizations) ;
    }

    if (optimizations.size() == 0)
    {
	osgUtil::StatsVisitor stats;
	if (dtkMsg.isSeverity(DTKMSG_INFO))
	{
	    dtkMsg.add(DTKMSG_INFO, "iris-convert: scenegraph statistics:\n") ;
	    node->accept(stats);
	    stats.totalUpStats();
	    stats.print(std::cerr) ;
	}
    }

    // write to the output file
    dtkMsg.add(DTKMSG_INFO, "iris-convert: writing %s\n",args[1]) ;
    if (!osgDB::writeNodeFile(*node, args[1]))
    {
	dtkMsg.add(DTKMSG_ERROR, "iris-convert: Unable to write to file \"%s\"\n", args[1]) ;
    }
    return 0 ;
}
